### Data Loading and Initial Inspection

In [ ]:
# Load the CSV file's data into a DataFrame named bank_loans
import json
import pandas as pd
import numpy as np
bank_loans = pd.read_csv("/Users/rohanchawla/Downloads/bank.txt", delimiter=",", skiprows=23)


In [ ]:
#  Use appropriate Python command(s) to inspect bank_loans 
bank_loans.info()     
bank_loans.head()       
bank_loans.shape       
bank_loans.columns    
bank_loans.describe()  

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   CustomerNo       418 non-null    int64 
 1   ApplicationDate  418 non-null    object
 2   LoanPurpose      418 non-null    object
 3   Checking         418 non-null    object
 4   Savings          418 non-null    object
 5   MonthsCustomer   418 non-null    int64 
 6   MonthsEmployed   418 non-null    int64 
 7   EducationLevel   418 non-null    object
 8   Gender           418 non-null    object
 9   MaritalStatus    418 non-null    object
 10  Age              418 non-null    int64 
 11  Housing          418 non-null    object
 12  Years            418 non-null    int64 
 13  Job              418 non-null    object
 14  CreditRisk       418 non-null    object
dtypes: int64(5), object(10)
memory usage: 49.1+ KB


CustomerNo  MonthsCustomer  MonthsEmployed         Age       Years
count  418.000000      418.000000      418.000000  418.000000  418.000000
mean   210.193780       22.894737       31.559809   34.342105    2.832536
std    121.875623       12.196765       32.254496   11.089788    1.084309
min      1.000000        5.000000        0.000000   18.000000    1.000000
25%    105.250000       13.000000        6.000000   26.000000    2.000000
50%    209.500000       19.000000       20.000000   32.000000    3.000000
75%    313.750000       28.000000       46.000000   41.000000    4.000000
max    425.000000       73.000000      119.000000   73.000000    4.000000

Write answers to questions here.

In [ ]:
bank_loans

CustomerNo ApplicationDate      LoanPurpose Checking  Savings  \
0             1        22/10/23          Repairs    $207       $0    
1             2         27/1/23          New Car    $193   $2,684    
2             3         21/8/23  Small Appliance      $0     $739    
3             4         10/1/23        Furniture      $0   $1,230    
4             5         11/4/23          New Car      $0     $389    
..          ...             ...              ...      ...      ...   
413         421          2/9/23        Furniture      $0     $957    
414         422         23/3/23          New Car      $0     $770    
415         423         14/3/23        Furniture    $983     $950    
416         424          9/1/23         Used Car      $0     $160    
417         425         31/3/23         Used Car      $0     $276    

     MonthsCustomer  MonthsEmployed     EducationLevel Gender MaritalStatus  \
0                28             116          Doctorate      M        Single   
1                13               5    Master's Degree      F      Divorced   
2                13              12   Associate Degree      M        Single   
3                25               0   Associate Degree      M      Divorced   
4                19             119        High School      M        Single   
..              ...             ...                ...    ...           ...   
413              19              11  Bachelor's Degree      F      Divorced   
414              37               3  Bachelor's Degree      F      Divorced   
415              13               5        High School      F      Divorced   
416              13               7    Master's Degree      M       Married   
417              25              91          Doctorate      M        Single   

     Age Housing  Years         Job CreditRisk  
0     47     Own      4     Skilled        Low  
1     22     Own      2   Unskilled       High  
2     23     Own      3   Unskilled        Low  
3     32     Own      1     Skilled       High  
4     38     Own      4  Management       High  
..   ...     ...    ...         ...        ...  
413   19    Rent      4     Skilled       High  
414   33     Own      4     Skilled       High  
415   24    Rent      3     Skilled       High  
416   40    Rent      4     Skilled        Low  
417   62     Own      4     Skilled        Low  

[418 rows x 15 columns]

1. There are 418 loan applications
2. There are total 15 columns. Columns that contain integer values are: CustomerNo, MonthsCustomer, MonthsEmployed,Age and Year. Columns that contain string values are ApplicationDate,LoanPurpose, Checking, Savings, EducationalLevel, Gender, MaritialStatus,Housing, Job, CreditRisk.

### Feature Engineering: Creating Total Balance

In [ ]:
# Add a new column called TotalBalance
bank_loans['Savings']=bank_loans['Savings'].str.strip("$").str.replace(",","").str.strip(" ").astype(int)

bank_loans['Checking']=bank_loans['Checking'].str.strip("$").str.replace(",","").str.strip(" ").astype(int)

bank_loans['TotalBalance']=bank_loans['Savings']+bank_loans['Checking']

bank_loans

CustomerNo ApplicationDate      LoanPurpose  Checking  Savings  \
0             1        22/10/23          Repairs       207        0   
1             2         27/1/23          New Car       193     2684   
2             3         21/8/23  Small Appliance         0      739   
3             4         10/1/23        Furniture         0     1230   
4             5         11/4/23          New Car         0      389   
..          ...             ...              ...       ...      ...   
413         421          2/9/23        Furniture         0      957   
414         422         23/3/23          New Car         0      770   
415         423         14/3/23        Furniture       983      950   
416         424          9/1/23         Used Car         0      160   
417         425         31/3/23         Used Car         0      276   

     MonthsCustomer  MonthsEmployed     EducationLevel Gender MaritalStatus  \
0                28             116          Doctorate      M        Single   
1                13               5    Master's Degree      F      Divorced   
2                13              12   Associate Degree      M        Single   
3                25               0   Associate Degree      M      Divorced   
4                19             119        High School      M        Single   
..              ...             ...                ...    ...           ...   
413              19              11  Bachelor's Degree      F      Divorced   
414              37               3  Bachelor's Degree      F      Divorced   
415              13               5        High School      F      Divorced   
416              13               7    Master's Degree      M       Married   
417              25              91          Doctorate      M        Single   

     Age Housing  Years         Job CreditRisk  TotalBalance  
0     47     Own      4     Skilled        Low           207  
1     22     Own      2   Unskilled       High          2877  
2     23     Own      3   Unskilled        Low           739  
3     32     Own      1     Skilled       High          1230  
4     38     Own      4  Management       High           389  
..   ...     ...    ...         ...        ...           ...  
413   19    Rent      4     Skilled       High           957  
414   33     Own      4     Skilled       High           770  
415   24    Rent      3     Skilled       High          1933  
416   40    Rent      4     Skilled        Low           160  
417   62     Own      4     Skilled        Low           276  

[418 rows x 16 columns]

In [ ]:
# List the customer information of the top 15 customers who have the lowest total balances
bank_loans.sort_values(by=['TotalBalance','CustomerNo'], ascending=[True,True]).head(15)


CustomerNo ApplicationDate      LoanPurpose  Checking  Savings  \
5             6         12/3/23          New Car         0        0   
25           26          4/5/23        Furniture         0        0   
29           30          3/2/23         Used Car         0        0   
58           59         10/8/23        Education         0        0   
73           74         18/3/23          New Car         0        0   
133         134         22/4/23        Furniture         0        0   
137         138          1/2/23            Other         0        0   
194         195         12/4/23          New Car         0        0   
285         286         19/8/23         Used Car         0        0   
290         291          3/2/23  Small Appliance         0        0   
403         411         22/6/23        Furniture         0        0   
410         418          3/5/23         Used Car         0        0   
319         320         28/5/23          New Car         0      102   
89           90        30/12/23         Used Car         0      104   
138         139         26/6/23         Business         0      104   

     MonthsCustomer  MonthsEmployed     EducationLevel Gender MaritalStatus  \
5                25             103   Associate Degree      F      Divorced   
25               25              23          Doctorate      M       Married   
29               19              58   Associate Degree      M        Single   
58               37             114          Doctorate      M        Single   
73               22               9    Master's Degree      M        Single   
133              13              94        High School      M        Single   
137              25              54  Bachelor's Degree      M        Single   
194              25              19    Master's Degree      F      Divorced   
285              37              49   Associate Degree      M        Single   
290              43              28        High School      F      Divorced   
403              40              30  Bachelor's Degree      M        Single   
410              11               4    Master's Degree      F      Divorced   
319               7               0        High School      F      Divorced   
89               37              25    Master's Degree      M        Single   
138              25              23  Bachelor's Degree      M       Married   

     Age Housing  Years         Job CreditRisk  TotalBalance  
5     28     Own      2     Skilled       High             0  
25    19     Own      4     Skilled       High             0  
29    50   Other      4     Skilled       High             0  
58    39     Own      4  Management       High             0  
73    39     Own      2   Unskilled       High             0  
133   48    Rent      4     Skilled        Low             0  
137   39     Own      3  Management       High             0  
194   24    Rent      4     Skilled       High             0  
285   46   Other      4     Skilled       High             0  
290   29     Own      3  Management       High             0  
403   29     Own      4  Management        Low             0  
410   30    Rent      4     Skilled        Low             0  
319   53     Own      4  Unemployed        Low           102  
89    23     Own      4     Skilled       High           104  
138   20     Own      2   Unskilled        Low           104

In [ ]:
bank_loans.sort_values(by=['TotalBalance','CustomerNo'], ascending=[True,True]).head(15).iloc[4]
bank_loans.sort_values(by=['TotalBalance','CustomerNo'], ascending=[True,True]).head(15)['TotalBalance'].iloc[14]-bank_loans.sort_values(by=['TotalBalance','CustomerNo'], ascending=[True,True]).head(15)['TotalBalance'].iloc[0]

np.int64(104)

Write answers to questions here.

Write answers to questions here.

Question 2

1. customer number for the fifth observation is CustomerNo 74.

2.  the value difference between the lowest and fifteenth-lowest total balances is 104.

### Temporal Analysis: Loan Applications by Month and Purpose

In [ ]:
# Create a new column called ApplicationMonth
bank_loans['ApplicationDate']=pd.to_datetime(bank_loans['ApplicationDate'])

bank_loans['ApplicationMonth']=bank_loans['ApplicationDate'].dt.month

bank_loans

/var/folders/_w/4nv5jqq10px66wg_9dn81bjw0000gn/T/ipykernel_10441/2656481543.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  bank_loans['ApplicationDate']=pd.to_datetime(bank_loans['ApplicationDate'])


CustomerNo ApplicationDate      LoanPurpose  Checking  Savings  \
0             1      2023-10-22          Repairs       207        0   
1             2      2023-01-27          New Car       193     2684   
2             3      2023-08-21  Small Appliance         0      739   
3             4      2023-10-01        Furniture         0     1230   
4             5      2023-11-04          New Car         0      389   
..          ...             ...              ...       ...      ...   
413         421      2023-02-09        Furniture         0      957   
414         422      2023-03-23          New Car         0      770   
415         423      2023-03-14        Furniture       983      950   
416         424      2023-09-01         Used Car         0      160   
417         425      2023-03-31         Used Car         0      276   

     MonthsCustomer  MonthsEmployed     EducationLevel Gender MaritalStatus  \
0                28             116          Doctorate      M        Single   
1                13               5    Master's Degree      F      Divorced   
2                13              12   Associate Degree      M        Single   
3                25               0   Associate Degree      M      Divorced   
4                19             119        High School      M        Single   
..              ...             ...                ...    ...           ...   
413              19              11  Bachelor's Degree      F      Divorced   
414              37               3  Bachelor's Degree      F      Divorced   
415              13               5        High School      F      Divorced   
416              13               7    Master's Degree      M       Married   
417              25              91          Doctorate      M        Single   

     Age Housing  Years         Job CreditRisk  TotalBalance  ApplicationMonth  
0     47     Own      4     Skilled        Low           207                10  
1     22     Own      2   Unskilled       High          2877                 1  
2     23     Own      3   Unskilled        Low           739                 8  
3     32     Own      1     Skilled       High          1230                10  
4     38     Own      4  Management       High           389                11  
..   ...     ...    ...         ...        ...           ...               ...  
413   19    Rent      4     Skilled       High           957                 2  
414   33     Own      4     Skilled       High           770                 3  
415   24    Rent      3     Skilled       High          1933                 3  
416   40    Rent      4     Skilled        Low           160                 9  
417   62     Own      4     Skilled        Low           276                 3  

[418 rows x 17 columns]

In [ ]:
# Determine the number of applications in each month
bank_loans.groupby('ApplicationMonth')['CustomerNo'].count()

ApplicationMonth
1     37
2     21
3     54
4     33
5     32
6     27
7     38
8     36
9     31
10    37
11    39
12    33
Name: CustomerNo, dtype: int64

In [ ]:
# Determine the number of applications for each loan purpose in each month
bank_loans.groupby(['ApplicationMonth','LoanPurpose'])['CustomerNo'].count()

ApplicationMonth  LoanPurpose    
1                 Business            6
                  Furniture           6
                  New Car            11
                  Other               3
                  Repairs             2
                                     ..
12                Large Appliance     1
                  New Car             8
                  Repairs             1
                  Small Appliance     7
                  Used Car            6
Name: CustomerNo, Length: 82, dtype: int64

In [ ]:
# Determine the number of different loan purposes in each month
bank_loans.groupby('ApplicationMonth')['LoanPurpose'].nunique()

ApplicationMonth
1     7
2     6
3     8
4     5
5     5
6     7
7     6
8     8
9     7
10    9
11    6
12    8
Name: LoanPurpose, dtype: int64

### Business Rule Check: Identifying Censored Loan Purposes

In [ ]:
# Create a function called check_cat_censored
def check_cat_censored(loan_purpose):
    censored_keywords=['Car','Appliance']
    for keyword in censored_keywords:
        if keyword in loan_purpose:
            return True
    else:
        return False

In [ ]:
# Create a new column called requires_censored
bank_loans['requires_censored']=bank_loans['LoanPurpose'].apply(check_cat_censored)

bank_loans

CustomerNo ApplicationDate      LoanPurpose  Checking  Savings  \
0             1      2023-10-22          Repairs       207        0   
1             2      2023-01-27          New Car       193     2684   
2             3      2023-08-21  Small Appliance         0      739   
3             4      2023-10-01        Furniture         0     1230   
4             5      2023-11-04          New Car         0      389   
..          ...             ...              ...       ...      ...   
413         421      2023-02-09        Furniture         0      957   
414         422      2023-03-23          New Car         0      770   
415         423      2023-03-14        Furniture       983      950   
416         424      2023-09-01         Used Car         0      160   
417         425      2023-03-31         Used Car         0      276   

     MonthsCustomer  MonthsEmployed     EducationLevel Gender MaritalStatus  \
0                28             116          Doctorate      M        Single   
1                13               5    Master's Degree      F      Divorced   
2                13              12   Associate Degree      M        Single   
3                25               0   Associate Degree      M      Divorced   
4                19             119        High School      M        Single   
..              ...             ...                ...    ...           ...   
413              19              11  Bachelor's Degree      F      Divorced   
414              37               3  Bachelor's Degree      F      Divorced   
415              13               5        High School      F      Divorced   
416              13               7    Master's Degree      M       Married   
417              25              91          Doctorate      M        Single   

     Age Housing  Years         Job CreditRisk  TotalBalance  \
0     47     Own      4     Skilled        Low           207   
1     22     Own      2   Unskilled       High          2877   
2     23     Own      3   Unskilled        Low           739   
3     32     Own      1     Skilled       High          1230   
4     38     Own      4  Management       High           389   
..   ...     ...    ...         ...        ...           ...   
413   19    Rent      4     Skilled       High           957   
414   33     Own      4     Skilled       High           770   
415   24    Rent      3     Skilled       High          1933   
416   40    Rent      4     Skilled        Low           160   
417   62     Own      4     Skilled        Low           276   

     ApplicationMonth  requires_censored  
0                  10              False  
1                   1               True  
2                   8               True  
3                  10              False  
4                  11               True  
..                ...                ...  
413                 2              False  
414                 3               True  
415                 3              False  
416                 9               True  
417                 3               True  

[418 rows x 18 columns]

In [ ]:
# Identify customers that satisfy the two conditions at the same time
filtered_customers=bank_loans.loc[(bank_loans['Years']>=3) & bank_loans['requires_censored']==True]
filtered_customers

CustomerNo ApplicationDate      LoanPurpose  Checking  Savings  \
2             3      2023-08-21  Small Appliance         0      739   
4             5      2023-11-04          New Car         0      389   
9            10      2023-11-21          New Car         0      229   
17           18      2023-08-25          New Car       396      228   
18           19      2023-07-18         Used Car         0      129   
..          ...             ...              ...       ...      ...   
410         418      2023-03-05         Used Car         0        0   
412         420      2023-01-10  Small Appliance       765    10406   
414         422      2023-03-23          New Car         0      770   
416         424      2023-09-01         Used Car         0      160   
417         425      2023-03-31         Used Car         0      276   

     MonthsCustomer  MonthsEmployed     EducationLevel Gender MaritalStatus  \
2                13              12   Associate Degree      M        Single   
4                19             119        High School      M        Single   
9                13              16   Associate Degree      M       Married   
17               13              26   Associate Degree      M        Single   
18               31               8    Master's Degree      M      Divorced   
..              ...             ...                ...    ...           ...   
410              11               4    Master's Degree      F      Divorced   
412              10              24  Bachelor's Degree      F      Divorced   
414              37               3  Bachelor's Degree      F      Divorced   
416              13               7    Master's Degree      M       Married   
417              25              91          Doctorate      M        Single   

     Age Housing  Years         Job CreditRisk  TotalBalance  \
2     23     Own      3   Unskilled        Low           739   
4     38     Own      4  Management       High           389   
9     26     Own      3   Unskilled        Low           229   
17    46     Own      3   Unskilled        Low           624   
18    39     Own      4  Management        Low           129   
..   ...     ...    ...         ...        ...           ...   
410   30    Rent      4     Skilled        Low             0   
412   65     Own      3   Unskilled        Low         11171   
414   33     Own      4     Skilled       High           770   
416   40    Rent      4     Skilled        Low           160   
417   62     Own      4     Skilled        Low           276   

     ApplicationMonth  requires_censored  
2                   8               True  
4                  11               True  
9                  11               True  
17                  8               True  
18                  7               True  
..                ...                ...  
410                 3               True  
412                 1               True  
414                 3               True  
416                 9               True  
417                 3               True  

[139 rows x 18 columns]

### Pivot Analysis: Education Level, Job Type, and Demographics

In [ ]:
# Generate a Pivot table to summarise the average age and the average number of years at current residence for different education levels and job types
pivot_summary = bank_loans.pivot_table(
    values=['Age', 'Years'],
    index=['EducationLevel', 'Job'],
    aggfunc='mean'
)

# Display the pivot table
print(pivot_summary)

                                    Age     Years
EducationLevel    Job                            
Associate Degree  Management  39.916667  3.000000
                  Skilled     32.938776  2.673469
                  Unemployed  25.000000  1.333333
                  Unskilled   29.333333  2.666667
Bachelor's Degree Management  37.312500  3.062500
                  Skilled     32.081633  2.673469
                  Unemployed  46.666667  3.333333
                  Unskilled   36.500000  2.666667
Doctorate         Management  33.500000  2.625000
                  Skilled     35.278689  2.852459
                  Unemployed  28.000000  4.000000
                  Unskilled   36.333333  2.833333
High School       Management  44.000000  2.888889
                  Skilled     32.125000  2.729167
                  Unemployed  53.000000  4.000000
                  Unskilled   35.823529  3.176471
Master's Degree   Management  38.000000  2.666667
                  Skilled     31.593220  3.033898


### Data Integration: Merging Loan Data with Monthly Information (XML)

In [ ]:
import sys
!{sys.executable} -m pip install lxml


[notice] A new release of pip is available: 25.0.1 -> 25.1.1
[notice] To update, run: pip3 install --upgrade pip


In [ ]:
# Load the XML file's data into a DataFrame named mon_info
mon_info=pd.read_xml("/Users/rohanchawla/Downloads/bank_extra.xml")
mon_info

Mon  MaxAppNum
0   Jan         10
1   Feb         12
2   Mar         12
3   Apr         12
4   May         15
5   Jun         15
6   Jul         15
7   Aug         15
8   Sep         12
9   Oct         12
10  Nov         12
11  Dec         10

In [ ]:
# Combine bank_loans and mon_info into one single DataFrame loans
mon_info['Month_Number']=pd.to_datetime(mon_info['Mon'],format='%b').dt.month

loans=pd.merge(bank_loans,mon_info, left_on='ApplicationMonth', right_on='Month_Number')

loans

CustomerNo ApplicationDate      LoanPurpose  Checking  Savings  \
0             1      2023-10-22          Repairs       207        0   
1             2      2023-01-27          New Car       193     2684   
2             3      2023-08-21  Small Appliance         0      739   
3             4      2023-10-01        Furniture         0     1230   
4             5      2023-11-04          New Car         0      389   
..          ...             ...              ...       ...      ...   
413         421      2023-02-09        Furniture         0      957   
414         422      2023-03-23          New Car         0      770   
415         423      2023-03-14        Furniture       983      950   
416         424      2023-09-01         Used Car         0      160   
417         425      2023-03-31         Used Car         0      276   

     MonthsCustomer  MonthsEmployed     EducationLevel Gender MaritalStatus  \
0                28             116          Doctorate      M        Single   
1                13               5    Master's Degree      F      Divorced   
2                13              12   Associate Degree      M        Single   
3                25               0   Associate Degree      M      Divorced   
4                19             119        High School      M        Single   
..              ...             ...                ...    ...           ...   
413              19              11  Bachelor's Degree      F      Divorced   
414              37               3  Bachelor's Degree      F      Divorced   
415              13               5        High School      F      Divorced   
416              13               7    Master's Degree      M       Married   
417              25              91          Doctorate      M        Single   

     ...  Housing Years         Job CreditRisk TotalBalance  ApplicationMonth  \
0    ...      Own     4     Skilled        Low          207                10   
1    ...      Own     2   Unskilled       High         2877                 1   
2    ...      Own     3   Unskilled        Low          739                 8   
3    ...      Own     1     Skilled       High         1230                10   
4    ...      Own     4  Management       High          389                11   
..   ...      ...   ...         ...        ...          ...               ...   
413  ...     Rent     4     Skilled       High          957                 2   
414  ...      Own     4     Skilled       High          770                 3   
415  ...     Rent     3     Skilled       High         1933                 3   
416  ...     Rent     4     Skilled        Low          160                 9   
417  ...      Own     4     Skilled        Low          276                 3   

     requires_censored  Mon MaxAppNum  Month_Number  
0                False  Oct        12            10  
1                 True  Jan        10             1  
2                 True  Aug        15             8  
3                False  Oct        12            10  
4                 True  Nov        12            11  
..                 ...  ...       ...           ...  
413              False  Feb        12             2  
414               True  Mar        12             3  
415              False  Mar        12             3  
416               True  Sep        12             9  
417               True  Mar        12             3  

[418 rows x 21 columns]

In [ ]:
# Sort loans by the column ApplicationMonth descendingly first and then by the customer number ascendingly. Then, save the sorted results in a new DataFrame sorted_loans.
sorted_loans=loans.sort_values(by=['ApplicationMonth','CustomerNo'],ascending=[False,True]).reset_index(drop=True)

sorted_loans

CustomerNo ApplicationDate      LoanPurpose  Checking  Savings  \
0             6      2023-12-03          New Car         0        0   
1            25      2023-12-09        Furniture       101     3871   
2            28      2023-12-15          New Car         0    10723   
3            37      2023-12-21         Business     16647      895   
4            51      2023-12-04          New Car         0      660   
..          ...             ...              ...       ...      ...   
413         381      2023-01-26        Furniture         0      435   
414         383      2023-01-05          New Car       135        0   
415         401      2023-01-01         Business       670     4014   
416         415      2023-01-31          New Car         0      547   
417         420      2023-01-10  Small Appliance       765    10406   

     MonthsCustomer  MonthsEmployed     EducationLevel Gender MaritalStatus  \
0                25             103   Associate Degree      F      Divorced   
1                13               5          Doctorate      F      Divorced   
2                11              15          Doctorate      M        Single   
3                16              34    Master's Degree      M        Single   
4                17              75   Associate Degree      M        Single   
..              ...             ...                ...    ...           ...   
413              19              16  Bachelor's Degree      F      Divorced   
414              37               7        High School      M        Single   
415              31              21        High School      F      Divorced   
416              13              40    Master's Degree      M      Divorced   
417              10              24  Bachelor's Degree      F      Divorced   

     ...  Housing Years        Job CreditRisk TotalBalance  ApplicationMonth  \
0    ...      Own     2    Skilled       High            0                12   
1    ...     Rent     4    Skilled       High         3972                12   
2    ...     Rent     2  Unskilled        Low        10723                12   
3    ...     Rent     4    Skilled        Low        17542                12   
4    ...     Rent     4    Skilled       High          660                12   
..   ...      ...   ...        ...        ...          ...               ...   
413  ...     Rent     4    Skilled       High          435                 1   
414  ...    Other     4    Skilled       High          135                 1   
415  ...     Rent     4  Unskilled       High         4684                 1   
416  ...      Own     3    Skilled       High          547                 1   
417  ...      Own     3  Unskilled        Low        11171                 1   

     requires_censored  Mon MaxAppNum  Month_Number  
0                 True  Dec        10            12  
1                False  Dec        10            12  
2                 True  Dec        10            12  
3                False  Dec        10            12  
4                 True  Dec        10            12  
..                 ...  ...       ...           ...  
413              False  Jan        10             1  
414               True  Jan        10             1  
415              False  Jan        10             1  
416               True  Jan        10             1  
417               True  Jan        10             1  

[418 rows x 21 columns]

### Data Integration: Merging Loan Data with JSON Information

In [ ]:
# Load the JSON file's data into a DataFrame named loan_info
loan_info=pd.read_json('/Users/rohanchawla/Downloads/bank_extra.json')

loan_info

Business  Education  Furniture  Large Appliance  New Car  \
MinProcessingTime         7          1          1                1        3   
MaxProcessingTime        14          7          7                7       14   

                   Other  Repairs  Retraining  Small Appliance  Used Car  
MinProcessingTime      1        1           1                1         3  
MaxProcessingTime     30        7           8                5        14

In [ ]:
# Combine sorted_loans and loan_info into one single DataFrame loans_final
loan_info_t = loan_info.T.reset_index()
loan_info_t

# Merge with sorted_loans
loans_final=pd.merge(sorted_loans,loan_info_t,left_on='LoanPurpose',right_on='index')
loans_final


CustomerNo ApplicationDate      LoanPurpose  Checking  Savings  \
0             6      2023-12-03          New Car         0        0   
1            25      2023-12-09        Furniture       101     3871   
2            28      2023-12-15          New Car         0    10723   
3            37      2023-12-21         Business     16647      895   
4            51      2023-12-04          New Car         0      660   
..          ...             ...              ...       ...      ...   
413         381      2023-01-26        Furniture         0      435   
414         383      2023-01-05          New Car       135        0   
415         401      2023-01-01         Business       670     4014   
416         415      2023-01-31          New Car         0      547   
417         420      2023-01-10  Small Appliance       765    10406   

     MonthsCustomer  MonthsEmployed     EducationLevel Gender MaritalStatus  \
0                25             103   Associate Degree      F      Divorced   
1                13               5          Doctorate      F      Divorced   
2                11              15          Doctorate      M        Single   
3                16              34    Master's Degree      M        Single   
4                17              75   Associate Degree      M        Single   
..              ...             ...                ...    ...           ...   
413              19              16  Bachelor's Degree      F      Divorced   
414              37               7        High School      M        Single   
415              31              21        High School      F      Divorced   
416              13              40    Master's Degree      M      Divorced   
417              10              24  Bachelor's Degree      F      Divorced   

     ...  CreditRisk TotalBalance  ApplicationMonth requires_censored  Mon  \
0    ...        High            0                12              True  Dec   
1    ...        High         3972                12             False  Dec   
2    ...         Low        10723                12              True  Dec   
3    ...         Low        17542                12             False  Dec   
4    ...        High          660                12              True  Dec   
..   ...         ...          ...               ...               ...  ...   
413  ...        High          435                 1             False  Jan   
414  ...        High          135                 1              True  Jan   
415  ...        High         4684                 1             False  Jan   
416  ...        High          547                 1              True  Jan   
417  ...         Low        11171                 1              True  Jan   

     MaxAppNum  Month_Number            index MinProcessingTime  \
0           10            12          New Car                 3   
1           10            12        Furniture                 1   
2           10            12          New Car                 3   
3           10            12         Business                 7   
4           10            12          New Car                 3   
..         ...           ...              ...               ...   
413         10             1        Furniture                 1   
414         10             1          New Car                 3   
415         10             1         Business                 7   
416         10             1          New Car                 3   
417         10             1  Small Appliance                 1   

     MaxProcessingTime  
0                   14  
1                    7  
2                   14  
3                   14  
4                   14  
..                 ...  
413                  7  
414                 14  
415                 14  
416                 14  
417                  5  

[418 rows x 24 columns]

In [ ]:
# Sort loans_final by the customer number descendingly and save the sorted results in a new DataFrame sorted_loans_final
sorted_loans_final=loans_final.sort_values(by='CustomerNo',ascending=False).reset_index(drop=True)
sorted_loans_final

CustomerNo ApplicationDate      LoanPurpose  Checking  Savings  \
0           425      2023-03-31         Used Car         0      276   
1           424      2023-09-01         Used Car         0      160   
2           423      2023-03-14        Furniture       983      950   
3           422      2023-03-23          New Car         0      770   
4           421      2023-02-09        Furniture         0      957   
..          ...             ...              ...       ...      ...   
413           5      2023-11-04          New Car         0      389   
414           4      2023-10-01        Furniture         0     1230   
415           3      2023-08-21  Small Appliance         0      739   
416           2      2023-01-27          New Car       193     2684   
417           1      2023-10-22          Repairs       207        0   

     MonthsCustomer  MonthsEmployed     EducationLevel Gender MaritalStatus  \
0                25              91          Doctorate      M        Single   
1                13               7    Master's Degree      M       Married   
2                13               5        High School      F      Divorced   
3                37               3  Bachelor's Degree      F      Divorced   
4                19              11  Bachelor's Degree      F      Divorced   
..              ...             ...                ...    ...           ...   
413              19             119        High School      M        Single   
414              25               0   Associate Degree      M      Divorced   
415              13              12   Associate Degree      M        Single   
416              13               5    Master's Degree      F      Divorced   
417              28             116          Doctorate      M        Single   

     ...  CreditRisk TotalBalance  ApplicationMonth requires_censored  Mon  \
0    ...         Low          276                 3              True  Mar   
1    ...         Low          160                 9              True  Sep   
2    ...        High         1933                 3             False  Mar   
3    ...        High          770                 3              True  Mar   
4    ...        High          957                 2             False  Feb   
..   ...         ...          ...               ...               ...  ...   
413  ...        High          389                11              True  Nov   
414  ...        High         1230                10             False  Oct   
415  ...         Low          739                 8              True  Aug   
416  ...        High         2877                 1              True  Jan   
417  ...         Low          207                10             False  Oct   

     MaxAppNum  Month_Number            index MinProcessingTime  \
0           12             3         Used Car                 3   
1           12             9         Used Car                 3   
2           12             3        Furniture                 1   
3           12             3          New Car                 3   
4           12             2        Furniture                 1   
..         ...           ...              ...               ...   
413         12            11          New Car                 3   
414         12            10        Furniture                 1   
415         15             8  Small Appliance                 1   
416         10             1          New Car                 3   
417         12            10          Repairs                 1   

     MaxProcessingTime  
0                   14  
1                   14  
2                    7  
3                   14  
4                    7  
..                 ...  
413                 14  
414                  7  
415                  5  
416                 14  
417                  7  

[418 rows x 24 columns]